pennylane/beta/queuing/operation.py

```diff
@@ -0,0 +1,1469 @@
+        """
+        return super().eigvals
+
+    def __init__(self, *params, wires=None, do_queue=True):
+        # extract the arguments
+        if wires is None:
+            wires = params[-1]
+            params = params[:-1]
+
+        super().__init__(*params, wires=wires, do_queue=do_queue)
+
+    def __repr__(self):
+        """Constructor-call-like representation."""
+        temp = super().__repr__()
+
+        if self.return_type is None:
+            return temp
+
+        if self.return_type is Probability:
+            return repr(self.return_type) + "(wires={})".format(self.wires.tolist())
+
+        return repr(self.return_type) + "(" + temp + ")"
+
+    def __matmul__(self, other):
+        if isinstance(other, Tensor):
+            return other.__rmatmul__(self)
+
+        if isinstance(other, Observable):
+            return Tensor(self, other)
+
+        raise ValueError("Can only perform tensor products between observables.")
+
+    def diagonalizing_gates(self):
+        r"""Returns the list of operations such that they
+        diagonalize the observable in the computational basis.
+
+        Returns:
+            list(qml.Operation): A list of gates that diagonalize
+            the observable in the computational basis.
+        """
+        raise NotImplementedError
+
+
+class Tensor(Observable):
+    """Container class representing tensor products of observables.
+
+    To create a tensor, simply initiate it like so:
+
+    >>> T = Tensor(qml.PauliX(0), qml.Hermitian(A, [1, 2]))
+
+    You can also create a tensor from other Tensors:
+
+    >>> T = Tensor(T, qml.PauliZ(4))
+
+    The ``@`` symbol can be used as a tensor product operation:
+
+    >>> T = qml.PauliX(0) @ qml.Hadamard(2)
+    """
+
+    # pylint: disable=abstract-method
+    return_type = None
+    tensor = True
+    par_domain = None
+
+    def __init__(self, *args):  # pylint: disable=super-init-not-called
+
+        self._eigvals_cache = None
+        self.obs = []
+
+        for o in args:
+            if isinstance(o, Tensor):
+                self.obs.extend(o.obs)
+            elif isinstance(o, Observable):
+                self.obs.append(o)
+            else:
+                raise ValueError("Can only perform tensor products between observables.")
+
+        self.queue()
+
+    def __str__(self):
+        """Print the tensor product and some information."""
+        return "Tensor product {}: {} params, wires {}".format(
+            [i.name for i in self.obs], len(self.data), self.wires.tolist()
+        )
+
+    def __repr__(self):
+        """Constructor-call-like representation."""
+        return "Tensor(" + ", ".join([repr(o) for o in self.obs]) + ")"
+
+    def queue(self):
+        qml.QueuingContext.append(self, owns=tuple(self.obs))
+
+        for o in self.obs:
+            try:
+                qml.QueuingContext.update_info(o, owner=self)
+            except AttributeError:
+                pass
+
+        return self
```

Comment:
This addition allows the use of `AnnotatedQueues` through the call to the `QueuingContext.update_info` method